src/web/server/htmlTemplate.ts

```diff
@@ -203,8 +204,29 @@ export const htmlTemplate = ({
                     <img src="https://sb.scorecardresearch.com/p?c1=2&c2=6035250&cv=2.0&cj=1&cs_ucfr=0&comscorekw=${keywords}" />
                 </noscript>
                 ${[...priorityScriptTags].join('\n')}
-                <style class="webfont">${getFontsCss()}${resetCSS}${css}</style>
-
+                <style data-auto-foft-fonts>${getFontsCss()}</style>
+                <script>
+                // mark: PswXqO - keep these in sync with preloads
+                window.autoFoft = {
+                    isCritical: function (font) {
+                        switch (font.family) {
+                            case 'GuardianTextEgyptian':
+                            case 'Guardian Text Egyptian Web':
+                            case 'GuardianTextSans':
+                            case 'Guardian Text Sans Web':
+                                return font.weight === 'normal' || font.weight === '400';
+                            case 'GH Guardian Headline':
+                            case 'Guardian Egyptian Web':
+                                return (font.weight === '500' || font.weight === '700') &&
```

Comment:
I think this will miss opinion articles for headlines, maybe that's ok? https://www.theguardian.com/commentisfree/2020/nov/27/brexit-festival-eu-national

Reply:
Yep. But here we enter the territory of having different defaults depending on where one lands (in their very first visit). If that could be automated…

Reply:
yeah it does – do we know about content type in htmlTemplate?

Reply:
We do, yeah, we can generate it in document.tsx from - I think we want to do it based on `designType` and `Display`, mirroring, at least for headline the logic here: https://github.com/guardian/dotcom-rendering/blob/main/src/web/components/ArticleHeadline.tsx#L176

What do you reckon @oliverlloyd ? Do we need a `decideFontStyle` mechanism?

Reply:
I wonder if that is a bit overkill currently? will there be more than the comment difference?

Reply:
although if that could pick critical fonts per content type it could also be used for preloads too, one place to define them

Reply:
This PR is a really clever idea! (Sorry, missed it earlier)

So I've added a function `isCritical` which has a big, wandering decision tree to decide if a particular font should be deemed critical or not depending on `designType` and `display`. It's just stand alone, I've not wired it in. I took the logic from browsing our [layout stories](https://www.chromatic.com/pullrequest?appId=5dfcbf3012392c0020e7140b&number=2138&view=components&groupPrefix=Layouts) (I used [David's branch](https://github.com/guardian/dotcom-rendering/pull/2138) to do this as that PR fixes a bug we had before where we weren't loading fonts in Storybook properly)

#### Types need reviewing
I noticed we have a bit of a typing clash where `font.weight` can be either a string or a number. Looks like our existing types expected a number but here we're sometimes checking for `font.weight === '400' || font.weight === 'normal'`. I've left of this check for now but this will need reviewing.

#### Verbose vs. Abstraction
I initially started to abstract the repeated code here but I pulled back because I've noticed where we've used this pattern elsewhere that keeping it verbose like this makes edits later on a lot easier and I can see this being an area where edits are likely to happen as we start to change and add more page layouts (we already have open tickets for this work).

Reply:
> critical or not depending on `designType` and `display`

Superb! That’s articles-only, right? I wonder if fronts differ enough to warrant having their default(s) too?

Reply:
I'm actually wondering about changing the API, so:

```html
<style data-auto-foft-fonts="0">
@font-face etc
</style>
<style data-auto-foft-fonts="100">
@font-face etc
</style>
```

then load each font found in a `data-auto-foft-fonts` stylesheet in order of the index (attribute value). allows you to customise declaratively rather than in script, and to create multiple batches if your page could benefit from it

Reply:
(and then prob not call it `auto-foft` any more ¯\\_(ツ)_/¯ )

Reply:
I wanted to return to this as I can see this being a useful feature to have when we adopt support for Labs articles, where an entirely different font is used. There would need to be a bit of work to bring this code up to date with the new format for `design` but that should be straight forward, happy to help with that.

@sndrs Apologies for the huge delay in responding. I didn't follow your proposed changes before and meant to ask some questions. Regarding the different api you proposed, how would that work with focussing on different fonts based on design type?

Reply:
I plan to revisit the API after this week, once I'm on devex, but I'm pretty certain it will very html-based, rather than react/dcr etc oriented

so I guess you'd have to move some decision making to the templating? applications would still need to work out which fonts they need and in what order, and provide them as `style` elements defining `@font-face`s for them

Reply:
> it will very html-based

Is this because we don't want to commit too much to React and keep a door open for a non react world in the future? If so, I can appreciate the benefit to this but I worry that having similar decision logic in two different places could be a pain.

We're doing a lot of work to [centralise these types](https://github.com/guardian/types/blob/main/src/format.ts) of decisions, [for colours](https://docs.google.com/document/d/1nkntFJUunqrwNYszTrc2RBfyYUd1bEAeQh76yVFOLjo/edit#), for layouts, [for design](https://github.com/guardian/dotcom-rendering/blob/fdac1eb8d5a31b237c90a1d334620cf5a0a36681/src/web/lib/decideDesignType.ts) and I'd ❤️ to add fonts to that process!

Of course, if it's for other reasons then 👍

Reply:
> Is this because we don't want to commit too much to React and keep a door open for a non react world in the future?

100% this. 

this is a mechanism for improving font loading anywhere, not for deciding which fonts a platform needs

the more those decisions are tied up in the implementation details of the platform, the more likely it is the platform will stagnate in the future, as it becomes harder and harder to port common behaviours to new implementations 

I appreciate it's easier to say that than do it, but to me DCR itself is proof of the value of rearchitecting parts of a platform in situe, and frontend font loading is an example of how limiting it becomes to bind performance enhancements to a single platform.

whatever happens with dotcom design and information architecture in the future, it's very probable we'll want to prioritise font resource loading, and that's all this is intended to do

Reply:
I'm picturing something like this in the document template

```js
// document.tsx

export default () =>
    getFontsForThisPageSomehow.map(
        (fontGroup, i) => `
            <style data-auto-foft-fonts=${i}>
            ${fontGroup.map(
                (font) => `
                @font-face {
                    font-family: ${font.family};
                    font-style: ${font.style};
                    font-weight: ${font.weight};
                    src: url(${font.files.woff2}) format('woff2');
                }
            `,
            )}
            </style>
            <script>
            ${fontLoadingJS}
            </script>`,
    )
```

how `getFontsForThisPageSomehow` would be totally DCR-specific – does that make sense?

Reply:
> this is a mechanism for improving font loading anywhere, not for deciding which fonts a platform needs

Gotcha. And thanks loads for the example, that really helps!

I think you're basically right, we should be thinking about where we want to be in 5 years time, not 6 months. Longer delays to rebuilds and making them smoother represent huge cost savings.

Doing this though is hard, but definately worth the effort, esp. in these higher level areas, like fonts.